### PR TITLE
TRUNK-4473 : Form Resource (Long-Text) causes error with binary file upload

### PR DESCRIPTION
**Pull request by karan173**
Fixes [TRUNK-4473](https://issues.openmrs.org/browse/TRUNK-4473)

### CODE REVIEW

File: web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java [hunk 1]
```diff
@@ -10,12 +10,15 @@
 package org.openmrs.web.controller.form;
 
 import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
 
 import org.apache.commons.lang.StringUtils;
+import org.jfree.util.Log;
```
Comment (djazayeri) [inline web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java]: This is wrong. Please see other OpenMRS code to see the standard way we log, by convention (e.g. it uses log4j, not jfree.util).

File: web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java [hunk 2]
```diff
@@ -81,7 +84,14 @@ public String handleAddFormResource(@ModelAttribute("resource") FormResource res
...
 			throw new RuntimeException("Error handling not yet implemented");
 		} else {
-			Context.getFormService().saveFormResource(resource);
+			try {
+				Context.getFormService().saveFormResource(resource);
+			}
+			catch (Exception ex) {
+				HttpSession httpSession = request.getSession();
+				Log.error("Error while adding form resource", ex);
+				httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "form.resource.add.failure");
```
Comment (djazayeri) [inline web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java]: Is there a way that we can give a more specific error message? (E.g. in the specific case reported in this ticket, someone tried to upload a binary file when only text is accepted. Is there any way we can tell that's the error and report it?)

**Comment by djazayeri** [comment]
Required: please log consistently with standard OpenMRS practice
Optional: is there a way we can give a more specific error message in some cases?

(You can reopen this PR after addressing those issues.)

**Comment by karan173** [comment]
@djazayeri Can you please refer me to some code where such error handling is done. The errors thrown are all related to the database layer. There are various kind of resources present and a no of errors are being thrown, for eg, when one or both fields are empty in the form. Hence, I am not able to extract meaningful info. from the exceptions thrown.

**Comment by djazayeri** [comment]
I suggest you look into whether it's possible to have the underlying code (in LongFreeTextDatatype or DatatypeServiceImpl) throw a meaningful exception in this specific case, or else report a more meaningful error message as a string message in the InvalidCustomValueException. It's probably better to display the message from there, rather than displaying _nothing_ to the user.

**Comment by karan173** [comment]
But the error is not localized to the LongFreeTextDatatype. There is no error handling and input validation for adding resources to a form. So even if we leave the resource name and/or value empty, uncaught exceptions are thrown. So should I just focus on the case described solely in this issue?

I was thinking of first adding validation for the presence of resource name and resource value form fields(since they can't be null). According to me we should only then think about catching exceptions specific to a datatype.

**Comment by karan173** [comment]
Another thing I have found is that, the handlers(say DateFieldGenDatatypeHandler) are not throwing InvalidCustomValueException for incompatible resource values because in the getValue method of WebAttributeUtils([link](https://github.com/openmrs/openmrs-core/blob/master/web/src/main/java/org/openmrs/web/attribute/WebAttributeUtil.java#L56)), a classCastException is being thrown, while typecasting the handler to its subclasses.

File: web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java (hunk 2 above)
Comment (karan173) [inline web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java]: @djazayeri  Do we wish to upload only files of the content type text/plain?

Comment (karan173) [inline web/src/main/java/org/openmrs/web/controller/form/FormResourceController.java]: @djazayeri should we only allow text/plain files or files like text/html as well?

**Pull request closed without merging**